lib/cfa/augeas_parser.rb

```diff
@@ -38,33 +41,41 @@ def add(value, placer = AppendPlacer.new)
       element = placer.new_element(@tree)
       element[:key] = augeas_name
       element[:value] = value
+      element[:operation] = :add
       # FIXME: load_collection missing here
     end
 
     def delete(value)
       key = augeas_name
-      @tree.data.reject! do |entry|
-        entry[:key] == key &&
-          if value.is_a?(Regexp)
-            value =~ entry[:value]
-          else
-            value == entry[:value]
-          end
+      to_remove = @tree.data.select do |entry|
+        entry[:key] == key && value_match?(entry[:value], value)
       end
 
+      to_remove.each { |e| e[:operation] = :remove }
+
       load_collection
     end
 
   private
 
     def load_collection
-      entries = @tree.data.select { |d| d[:key] == augeas_name }
+      entries = @tree.data.select do |entry|
+        entry[:key] == augeas_name && entry[:operation] != :remove
+      end
       @collection = entries.map { |e| e[:value] }.freeze
     end
 
     def augeas_name
       @name + "[]"
     end
+
+    def value_match?(value, match)
```

Comment:
there was some problem in past, do not remember reason, but I am changing it back. I think rubocop complain.